### PR TITLE
Fix the schema mismatch between logical and physical for aggregate function, add `AggregateUDFImpl::is_null`

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.
Part of #11782 , it would be nice to cleanup schema before fighting with physical name

### CODE REVIEW

File: datafusion/core/src/physical_planner.rs [hunk 1]
```diff
@@ -670,6 +670,10 @@ impl DefaultPhysicalPlanner {
                 let input_exec = children.one()?;
                 let physical_input_schema = input_exec.schema();
                 let logical_input_schema = input.as_ref().schema();
+                let physical_input_schema_from_logical: Arc<Schema> =
+                    logical_input_schema.as_ref().clone().into();
+
+                debug_assert_eq!(physical_input_schema_from_logical, physical_input_schema, "Physical input schema should be the same as the one converted from logical input schema. Please file an issue or send the PR");
```
Comment (jayzhan211) [inline datafusion/core/src/physical_planner.rs:676]: The main goal of the change is to ensure they are the same.

File: datafusion/core/src/physical_planner.rs [hunk 2]
```diff
@@ -1599,11 +1603,10 @@ pub fn create_aggregate_expr_with_name_and_maybe_filter(
                 let ordering_reqs: Vec<PhysicalSortExpr> =
                     physical_sort_exprs.clone().unwrap_or(vec![]);
 
-                let schema: Schema = logical_input_schema.clone().into();
```
Comment (jayzhan211) [inline datafusion/core/src/physical_planner.rs:1602]: workaround cleanup

File: datafusion/expr/src/expr_schema.rs [hunk 3]
```diff
@@ -328,10 +328,45 @@ impl ExprSchemable for Expr {
...
+                        {
+                            Ok(false)
+                        } else {
+                            Ok(true)
+                        }
+                    }
+                    WindowFunctionDefinition::AggregateUDF(func) => {
+                        // TODO: UDF should be able to customize nullability
+                        if func.name() == "count" {
+                            // TODO: there is issue unsolved for count with window, should return false
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_schema.rs:347]: Not so familiar with window function yet, leave it as TODO

File: datafusion/physical-plan/src/windows/utils.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow_schema::{Schema, SchemaBuilder};
+use datafusion_common::Result;
+use datafusion_physical_expr::window::WindowExpr;
+use std::sync::Arc;
+
+pub(crate) fn create_schema(
```
Comment (jayzhan211) [inline datafusion/physical-plan/src/windows/utils.rs:23]: move the common function to utils. The logic is the same

File: datafusion/core/src/physical_planner.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/physical_planner.rs:676]: Nice!

Did you consider making this function return an `internal_error` rather than debug_assert ? 

If we are concerned about breaking existing tests, we could add a config setting like `datafusion.optimizer.skip_failed_rules` to let users bypass the check

File: datafusion/expr/src/expr_schema.rs (hunk 3 above)
Comment (alamb) [reply datafusion/expr/src/expr_schema.rs:347]: Perhaps we can file a ticket to track this -- ideally it would eventually be part of the window function definition itself rather than relying on names

File: datafusion/expr/src/expr_schema.rs [hunk 5]
```diff
@@ -328,10 +328,45 @@ impl ExprSchemable for Expr {
                     Ok(true)
                 }
             }
+            Expr::WindowFunction(WindowFunction { fun, .. }) => {
```
Comment (alamb) [inline datafusion/expr/src/expr_schema.rs:331]: Is this change required for this PR or is it a  "drive by" improvement?

File: datafusion/expr/src/expr_schema.rs [hunk 6]
```diff
@@ -328,10 +328,45 @@ impl ExprSchemable for Expr {
...
+                            Ok(true)
+                        } else {
+                            Ok(true)
+                        }
+                    }
+                    _ => Ok(true),
+                }
+            }
+            Expr::ScalarFunction(ScalarFunction { func, args }) => {
+                // If all the element in coalesce is non-null, the result is non-null
```
Comment (alamb) [inline datafusion/expr/src/expr_schema.rs:357]: We should probably add an API to ScalarUDFImpl to signal its null/non-nullness (as a follow on PR) instead of hard coding this function name

```
     func.is_nullable(args)
```

File: datafusion/expr/src/udaf.rs [hunk 7]
```diff
@@ -196,6 +196,10 @@ impl AggregateUDF {
         self.inner.state_fields(args)
     }
 
+    pub fn fields(&self, args: StateFieldsArgs) -> Result<Field> {
```
Comment (alamb) [inline datafusion/expr/src/udaf.rs:199]: Could we document this function and what it is for (also in AggregateUdfImpl)? 

Also, the name is strange to me -- it is `fields` but it returns a single `Field` and the corresponding function on `AggregateUDFImpl` is called `field` (no `s`) 🤔

File: datafusion/functions-aggregate-common/src/aggregate.rs [hunk 8]
```diff
@@ -171,6 +171,9 @@ pub trait AggregateExpr: Send + Sync + Debug + PartialEq<dyn Any> {
     fn get_minmax_desc(&self) -> Option<(Field, bool)> {
         None
     }
+
+    /// Get function's name, for example `count(x)` returns `count`
+    fn func_name(&self) -> &str;
```
Comment (alamb) [inline datafusion/functions-aggregate-common/src/aggregate.rs:176]: is there a reason this isn't `name()` ? `func_name` is fine, it just seems inconsistent with the rest of the code

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 9]
```diff
@@ -833,39 +841,47 @@ fn coerce_union_schema(inputs: &[Arc<LogicalPlan>]) -> Result<DFSchema> {
...
-                            plan_field.data_type(),
-                            plan_field.name()
-                        )
-                    },
-                )?;
-            *union_datatype = coerced_type;
-            *union_nullable = *union_nullable || plan_field.is_nullable();
+
+        // Safety: Length is checked
+        unsafe {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:846]: I think this unsafe block is  unecessary -- this isn't a performance critical piece of code. I think `izip` or just manuallly `zip`ping three times would be better

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 10]
```diff
@@ -80,6 +80,14 @@ impl WindowExpr for PlainAggregateWindowExpr {
     }
 
     fn field(&self) -> Result<Field> {
+        // TODO: Fix window function to always return non-null for count
```
Comment (alamb) [inline datafusion/physical-expr/src/window/aggregate.rs:83]: I don't understand this comment -- can we please file a ticket to track it (and add the ticket reference to the comments)?

File: datafusion/physical-expr/src/window/built_in.rs [hunk 11]
```diff
@@ -97,6 +97,10 @@ impl BuiltInWindowExpr {
 }
 
 impl WindowExpr for BuiltInWindowExpr {
+    fn func_name(&self) -> Result<&str> {
+        not_impl_err!("function name not determined")
```
Comment (alamb) [inline datafusion/physical-expr/src/window/built_in.rs:101]: why wouldn't we implement func_name for a built in window function 🤔

**Review by alamb — COMMENTED** [review]
This seems like a good change to me, but I don't fully understand how it is all connected . Thank you for taking this on @jayzhan211 

I am quite concerned about the use of `unsafe` but otherwise I think all this PR needs is  some TODOs with ticket references  and it would be good to go from my perspective.

File: datafusion/expr/src/expr_schema.rs (hunk 5 above)
Comment (jayzhan211) [reply datafusion/expr/src/expr_schema.rs:331]: Required

File: datafusion/functions-aggregate-common/src/aggregate.rs (hunk 8 above)
Comment (jayzhan211) [reply datafusion/functions-aggregate-common/src/aggregate.rs:176]: This is to identify function, there is `name()` already, but it includes arguments, which is not I want.
Alternative way is introduce `nullable()` for AggregateUDF, so we don't need name checking. Maybe I should done it before this PR.

File: datafusion/core/src/physical_planner.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/core/src/physical_planner.rs:676]: The change here is to ensure they are equivalent, thus I don't think we should let user bypass the check 🤔

File: datafusion/physical-expr/src/window/built_in.rs (hunk 11 above)
Comment (jayzhan211) [reply datafusion/physical-expr/src/window/built_in.rs:101]: The reason is because I don't need it -- for name checking in `nullable`

**Comment by jayzhan211** [comment]
I think `func_name` is indeed another workaround, would like to getting rid of it before this PR merged

File: datafusion/sqllogictest/test_files/count_star_rule.slt [hunk 12]
```diff
@@ -85,7 +85,7 @@ logical_plan
 03)----TableScan: t1 projection=[a]
 physical_plan
 01)ProjectionExec: expr=[a@0 as a, count() PARTITION BY [t1.a] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING@1 as count_a]
-02)--WindowAggExec: wdw=[count() PARTITION BY [t1.a] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING: Ok(Field { name: "count() PARTITION BY [t1.a] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(NULL)), end_bound: Following(UInt64(NULL)), is_causal: false }]
+02)--WindowAggExec: wdw=[count() PARTITION BY [t1.a] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING: Ok(Field { name: "count() PARTITION BY [t1.a] ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(NULL)), end_bound: Following(UInt64(NULL)), is_causal: false }]
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/count_star_rule.slt:88]: true -> false for count function

File: datafusion/sqllogictest/test_files/union.slt [hunk 13]
```diff
@@ -150,6 +150,21 @@ GROUP BY c1
...
 3 3
 
+# This test goes through schema check in aggregate plan
+query II rowsort
+SELECT c1, SUM(c2) FROM (
+    SELECT 1 as c1, 1::int as c2
+    UNION
+    SELECT 2 as c1, 2::int as c2
+    UNION
+    SELECT 3 as c1, count(1) as c2
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/union.slt:160]: if count's nullable is not matched, this test failed

File: datafusion/sqllogictest/test_files/window.slt [hunk 14]
```diff
@@ -1310,7 +1310,7 @@ logical_plan
...
-02)--BoundedWindowAggExec: wdw=[count(*) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING: Ok(Field { name: "count(*) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)), is_causal: false }], mode=[Sorted]
+02)--BoundedWindowAggExec: wdw=[count(*) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING: Ok(Field { name: "count(*) PARTITION BY [aggregate_test_100.c1] ORDER BY [aggregate_test_100.c2 ASC NULLS LAST] ROWS BETWEEN 1 PRECEDING AND 1 FOLLOWING", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Rows, start_bound: Preceding(UInt64(1)), end_bound: Following(UInt64(1)), is_causal: false }], mode=[Sorted]
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/window.slt:1313]: this file also contains count's true->false change only

File: datafusion/physical-expr/src/window/aggregate.rs [hunk 15]
```diff
@@ -176,9 +176,9 @@ impl AggregateWindowExpr for PlainAggregateWindowExpr {
         value_slice: &[ArrayRef],
         accumulator: &mut Box<dyn Accumulator>,
     ) -> Result<ScalarValue> {
-        let value = if cur_range.start == cur_range.end {
-            // We produce None if the window is empty.
-            ScalarValue::try_from(self.aggregate.field()?.data_type())?
+        if cur_range.start == cur_range.end {
+            self.aggregate
+                .default_value(self.aggregate.field()?.data_type())
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/window/aggregate.rs:181]: Find the root cause of window count function, so `func_name` is no longer required.

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 16]
```diff
@@ -836,17 +836,17 @@ mod tests {
...
+        let expected_input = ["BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow, is_causal: false }], mode=[Sorted]",
             "  FilterExec: NOT non_nullable_col@1",
             "    SortExec: expr=[non_nullable_col@1 ASC NULLS LAST], preserve_partitioning=[false]",
-            "      BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow, is_causal: false }], mode=[Sorted]",
+            "      BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow, is_causal: false }], mode=[Sorted]",
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:842]: count() is always non nullable, so this change makes sense to mee

File: datafusion/core/src/physical_planner.rs [hunk 17]
```diff
@@ -670,6 +670,12 @@ impl DefaultPhysicalPlanner {
                 let input_exec = children.one()?;
                 let physical_input_schema = input_exec.schema();
                 let logical_input_schema = input.as_ref().schema();
+                let physical_input_schema_from_logical: Arc<Schema> =
+                    logical_input_schema.as_ref().clone().into();
+
+                if physical_input_schema != physical_input_schema_from_logical {
```
Comment (alamb) [inline datafusion/core/src/physical_planner.rs:676]: 🎉

File: datafusion/expr/src/expr_schema.rs [hunk 18]
```diff
@@ -320,18 +320,28 @@ impl ExprSchemable for Expr {
...
+                        || func.name() == "NTILE"
+                        || func.name() == "CUME_DIST"
+                    {
+                        Ok(false)
+                    } else {
+                        Ok(true)
+                    }
+                }
+                WindowFunctionDefinition::AggregateUDF(func) => Ok(func.is_nullable()),
+                WindowFunctionDefinition::WindowUDF(udwf) => Ok(udwf.nullable()),
```
Comment (alamb) [inline datafusion/expr/src/expr_schema.rs:341]: this is great

File: datafusion/expr/src/expr_schema.rs [hunk 19]
```diff
@@ -320,18 +320,28 @@ impl ExprSchemable for Expr {
...
-                if func.name() == "count" {
-                    Ok(false)
-                } else {
-                    Ok(true)
-                }
+                Ok(func.is_nullable())
             }
+            Expr::WindowFunction(WindowFunction { fun, .. }) => match fun {
+                WindowFunctionDefinition::BuiltInWindowFunction(func) => {
+                    if func.name() == "RANK"
```
Comment (alamb) [inline datafusion/expr/src/expr_schema.rs:331]: checking the name is probably fine given we are in the process of removing the enum anyways

This could also check the variants of `func` here as well

File: datafusion/expr/src/udaf.rs [hunk 20]
```diff
@@ -342,6 +351,11 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
     /// the arguments
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType>;
 
+    /// Whether the aggregate function is nullable
```
Comment (alamb) [inline datafusion/expr/src/udaf.rs:354]:
```suggestion
    /// Whether the aggregate function is nullable.
    ///
    /// Nullable means that that the function could return `null` for any inputs.
    /// For example, aggregate functions like `COUNT` always return a non null value
    /// but others like `MIN` will return `NULL` if there is no non null input.
```

File: datafusion/expr/src/udaf.rs [hunk 21]
```diff
@@ -552,6 +566,13 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
     fn is_descending(&self) -> Option<bool> {
         None
     }
+
+    /// Returns default value of the function given the input is Null
```
Comment (alamb) [inline datafusion/expr/src/udaf.rs:570]:
```suggestion
    /// Returns default value of the function given the input is all `null`.
    ///
```

File: datafusion/functions-aggregate/src/count.rs [hunk 22]
```diff
@@ -121,6 +121,10 @@ impl AggregateUDFImpl for Count {
         Ok(DataType::Int64)
     }
 
+    fn is_nullable(&self) -> bool {
```
Comment (alamb) [inline datafusion/functions-aggregate/src/count.rs:124]: 👍

**Review by alamb — APPROVED** [review]
This looks like a really nice improvement to me -- thank you @jayzhan211 

cc @mustafasrepo  / @ozankabak  as I think you may be more familiar with the use of aggregate functions as window functions and might have additional thoughts you want to share

File: datafusion/functions/src/core/arrow_cast.rs [hunk 23]
```diff
@@ -87,6 +89,10 @@ impl ScalarUDFImpl for ArrowCastFunc {
         internal_err!("arrow_cast should return type from exprs")
     }
 
+    fn is_nullable(&self, args: &[Expr], schema: &dyn ExprSchema) -> bool {
+        args.iter().any(|e| e.nullable(schema).ok().unwrap_or(true))
```
Comment (jayzhan211) [inline datafusion/functions/src/core/arrow_cast.rs:93]: This was added due to the test from #12050

Comment (jayzhan211) [reply datafusion/functions/src/core/arrow_cast.rs:93]: It seems there are more functions required to customize nullable based on input's nullability. But we can add them gradually

File: datafusion/expr/src/expr_schema.rs [hunk 24]
```diff
@@ -335,18 +335,28 @@ impl ExprSchemable for Expr {
...
+                        || func.name() == "NTILE"
+                        || func.name() == "CUME_DIST"
```
Comment (berkaysynnada) [inline datafusion/expr/src/expr_schema.rs:348]: I am not sure if this list is complete. What about `DenseRank` and `PercentRank`?

File: datafusion/functions-aggregate-common/src/aggregate.rs [hunk 25]
```diff
@@ -171,6 +170,11 @@ pub trait AggregateExpr: Send + Sync + Debug + PartialEq<dyn Any> {
     fn get_minmax_desc(&self) -> Option<(Field, bool)> {
         None
     }
+
+    /// Returns default value of the function given the input is Null
+    /// Most of the aggregate function return Null if input is Null,
+    /// while `count` returns 0 if input is Null
+    fn default_value(&self, data_type: &DataType) -> Result<ScalarValue>;
```
Comment (berkaysynnada) [inline datafusion/functions-aggregate-common/src/aggregate.rs:177]: 👍🏻

File: datafusion/expr/src/udaf.rs [hunk 26]
```diff
@@ -552,6 +570,14 @@ pub trait AggregateUDFImpl: Debug + Send + Sync {
...
     }
+
+    /// Returns default value of the function given the input is all `null`.
+    ///
+    /// Most of the aggregate function return Null if input is Null,
+    /// while `count` returns 0 if input is Null
+    fn default_value(&self, data_type: &DataType) -> Result<ScalarValue> {
+        ScalarValue::try_from(data_type)
+    }
 }
```
Comment (berkaysynnada) [inline datafusion/expr/src/udaf.rs:581]: If an `AggregateUDFImpl ` overrides `is_nullable()` to return false but does not set the `default_value()`, it seems that it would indicate it is not nullable, yet its `default_value()` would return `null`. Is there a way to prevent this behavior? 🤔

File: datafusion/physical-plan/src/windows/window_agg_exec.rs [hunk 27]
```diff
@@ -33,19 +29,21 @@ use crate::{
...
 use datafusion_common::utils::{evaluate_partition_ranges, transpose};
 use datafusion_common::{internal_err, Result};
 use datafusion_execution::TaskContext;
 use datafusion_physical_expr::PhysicalSortRequirement;
-
 use futures::{ready, Stream, StreamExt};
+use std::any::Any;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
```
Comment (berkaysynnada) [inline datafusion/physical-plan/src/windows/window_agg_exec.rs:46]: Can you move them back to the top?

File: datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs [hunk 28]
```diff
@@ -59,13 +53,17 @@ use datafusion_physical_expr::window::{
...
 use futures::{ready, StreamExt};
 use hashbrown::raw::RawTable;
 use indexmap::IndexMap;
 use log::debug;
+use std::any::Any;
+use std::cmp::{min, Ordering};
+use std::collections::{HashMap, VecDeque};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
```
Comment (berkaysynnada) [inline datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs:67]: Can you move them back to top?

File: datafusion/physical-expr/src/window/built_in.rs [hunk 29]
```diff
@@ -35,6 +30,9 @@ use datafusion_common::utils::evaluate_partition_ranges;
 use datafusion_common::{Result, ScalarValue};
 use datafusion_expr::window_state::{WindowAggState, WindowFrameContext};
 use datafusion_expr::WindowFrame;
+use std::any::Any;
+use std::ops::Range;
+use std::sync::Arc;
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/window/built_in.rs:36]: Can you move them back to the top?

File: datafusion/expr/src/udaf.rs (hunk 26 above)
Comment (jayzhan211) [reply datafusion/expr/src/udaf.rs:581]: In this case, it should at least failed when creating record batch. Since the schema indicates non-null but got null value. 

If you are saying prevent `panic`, it is a good question, I don't have answer

File: datafusion/expr/src/expr_schema.rs (hunk 24 above)
Comment (jayzhan211) [reply datafusion/expr/src/expr_schema.rs:348]: Temporary code, there would be no name checking after #8709 is done. We can see that `Row_Number` is gone

File: datafusion/physical-plan/src/windows/window_agg_exec.rs (hunk 27 above)
Comment (jayzhan211) [reply datafusion/physical-plan/src/windows/window_agg_exec.rs:46]: I would appreciate a lot if there is `rustfmt.toml` so we just need `cargo fmt` and the import style is consistent

Comment (ozankabak) [reply datafusion/physical-plan/src/windows/window_agg_exec.rs:46]: I agree. We should automate this

Comment (alamb) [reply datafusion/physical-plan/src/windows/window_agg_exec.rs:46]: Yeah, I agree as well -- if we care about a consistent location / import order / style of `use` statements, let's get it in some automated tool that can be run as part of CI rather than try to enforce it manually during code review

I often see PRs that rearrange the imports and I suspect it is some IDE setting. Having it inconsistent makes merge conflicts more common but otherwise is not a problem in my mind

File: datafusion/expr/src/udaf.rs (hunk 26 above)
Comment (alamb) [reply datafusion/expr/src/udaf.rs:581]: Maybe we can improve the docuemention?

**Review by alamb — APPROVED** [review]

**Comment by jayzhan211** [comment]
Thanks @alamb @berkaysynnada @ozankabak

**Comment by phillipleblanc** [comment]
@jayzhan211 I'm in the process of upgrading spiceai to use DataFusion 42 and I'm running into the schema mismatch error from this PR:

`Internal("Physical input schema should be the same as the one converted from logical input schema.")`

I have a custom TableProvider and I get the error when running a `SELECT COUNT(1) FROM my_table`. This is what the explain plan looked like on DF 41:

```rust
let expected_plan = [
        "+---------------+--------------------------------------------------------------------------------+",
        "| plan_type     | plan                                                                           |",
        "+---------------+--------------------------------------------------------------------------------+",
        "| logical_plan  | Aggregate: groupBy=[[]], aggr=[[count(Int64(1))]]                              |",
        "|               |   BytesProcessedNode                                                           |",
        "|               |     TableScan: non_federated_abc projection=[]                                 |",
        "| physical_plan | AggregateExec: mode=Final, gby=[], aggr=[count(Int64(1))]                      |",
        "|               |   CoalescePartitionsExec                                                       |",
        "|               |     AggregateExec: mode=Partial, gby=[], aggr=[count(Int64(1))]                |",
        "|               |       BytesProcessedExec                                                       |",
        "|               |         SchemaCastScanExec                                                     |",
        "|               |           RepartitionExec: partitioning=RoundRobinBatch(3), input_partitions=1 |",
        "|               |             SqlExec sql=SELECT \"id\", \"created_at\" FROM non_federated_abc       |",
        "|               |                                                                                |",
        "+---------------+--------------------------------------------------------------------------------+",
    ];
```

(`BytesProcessedNode` and `BytesProcessedExec` are custom operators that we inject for tracking the number of bytes processed, I don't think its relevant to this bug - but I initially had a similar schema check for it that I ended up removing for the reason below)

My assumption of what is going on here is that logically no columns are required for the logical plan to come up with the count of the number of rows, but the TableProvider has to return all of the columns because it needs the rows to perform the count aggregation. But it ends up throwing away the columns because they get erased in the aggregation. Thus the check that the physical schema and the logical schema are equal is not strictly needed for this plan. Does that sound right?

**Comment by alamb** [comment]
@itsjunetime  is also having some issues related to this ticket during our upgrade of DataFusion, I am not sure if they are releated

**Comment by jayzhan211** [comment]
What is the logical schema and physical schema you have (the error)?

I think they should be consistent even for count (*) statement. They (logical & physical)should either have all the columns or no columns

**Comment by phillipleblanc** [comment]
I added `dbg!` statements to both, and I see one has the fields and the one from the logical plan is empty:

```console
[/Users/phillip/code/apache/datafusion/datafusion/core/src/physical_planner.rs:676:21] &physical_input_schema = Schema {
    fields: [
        Field {
            name: "id",
            data_type: Utf8,
            nullable: true,
            dict_id: 0,
            dict_is_ordered: false,
            metadata: {},
        },
        Field {
            name: "created_at",
            data_type: Timestamp(
                Nanosecond,
                Some(
                    "+00:00",
                ),
            ),
            nullable: true,
            dict_id: 0,
            dict_is_ordered: false,
            metadata: {},
        },
    ],
    metadata: {},
}
[/Users/phillip/code/apache/datafusion/datafusion/core/src/physical_planner.rs:677:21] &physical_input_schema_from_logical = Schema {
    fields: [],
    metadata: {},
}
thread 'acceleration::query_push_down::acceleration_with_and_without_federation' panicked at crates/runtime/tests/acceleration/query_push_down.rs:218:10:
collect working: Internal("Physical input schema should be the same as the one converted from logical input schema.")
```

**Comment by jayzhan211** [comment]
I think the ideal way is to have something like `wildcard field` for both logical and physical 🤔

**Comment by phillipleblanc** [comment]
> Could we keep all the fields for the logical plan to make them consistent?

That seems fine to me.

**Comment by itsjunetime** [comment]
> What is the logical schema and physical schema you have (the error)?

I'm getting issues where both schemas are equivalent except for the metadata on the fields. I think (correct me if I'm wrong) that field metadata doesn't actually need to be equivalent for the invariant that this error is trying to catch to be upheld. I think we could comfortably switch from directly comparing the schemas with `PartialEq` to using [`equivalent_names_and_types`](https://github.com/apache/datafusion/blob/98e5f64df37b7a3a70ce5018d0d4098d0bed6ddb/datafusion/common/src/dfschema.rs#L666) or something like that. I can file a ticket and reproducer/fix for this.

I'm also seeing issues that don't propagate with this exact error (but rather complain that `count(*)` is a non-nullable column that contains null values), so once I've figured out more specifics for this issue I can file another ticket for it.

**Comment by jayzhan211** [comment]
> I think (correct me if I'm wrong) that field metadata doesn't actually need to be equivalent for the invariant that this error is trying to catch to be upheld

If the metadata is mismatched, it indicates we lost the metadata somewhere when passing through the schema info, so I think it makes sense to check the metadata too. Maybe we should figure out the reason why metadata is mismatched first

**Comment by findepi** [comment]
thank you!
